zio-http/src/main/scala/zio/http/netty/NettyResponseEncoder.scala

```diff
@@ -0,0 +1,56 @@
+package zio.http.netty
+
+import io.netty.buffer.Unpooled
+import io.netty.handler.codec.http._
+import zio._
+import zio.http._
+
+import java.util.concurrent.ConcurrentHashMap
+
+private[zio] object NettyResponseEncoder {
+
+  private val frozenCache = new ConcurrentHashMap[Response, HttpResponse]()
+
+  def encode(response: Response): ZIO[Any, Throwable, HttpResponse] = {
+    val body = response.body
+    if (body.isComplete)
```

Comment:
Even in the cached case, you are executing `body.asArray.flatMap(...)`, which is a lot of unnecessary computation.

Instead, just check to see if the `frozenCache` has the response in it, and if it does, return that (pre-computed) effect directly.

Note that `Response#hashCode` and `Response#equals` will need to be very fast for such frozen path optimization to not interfere with the normal and common dynamic response case.